File: kora-java-graalvm-crud-r2dbc/src/test/java/ru/tinkoff/kora/example/graalvm/crud/r2dbc/AppContainer.java
package ru.tinkoff.kora.example.graalvm.crud.r2dbc;

import java.net.URI;
import java.nio.file.Paths;
import org.slf4j.LoggerFactory;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.output.Slf4jLogConsumer;
import org.testcontainers.containers.wait.strategy.Wait;
import org.testcontainers.images.builder.ImageFromDockerfile;
import org.testcontainers.utility.DockerImageName;

public final class AppContainer extends GenericContainer<AppContainer> {

    private AppContainer() {
        super(new ImageFromDockerfile("kora-java-graalvm-crud-r2dbc")
                .withDockerfile(Paths.get("Dockerfile").toAbsolutePath()));
    }

    private AppContainer(DockerImageName image) {
        super(image);
    }

    public static AppContainer build() {
        final String appImage = System.getenv("IMAGE_KORA_JAVA_GRAALVM_CRUD_R2DBC");
        return (appImage != null && !appImage.isBlank())
                ? new AppContainer(DockerImageName.parse(appImage))
                : new AppContainer();
    }

    @Override
    protected void configure() {
        super.configure();
        withExposedPorts(8080, 8085);
        withLogConsumer(new Slf4jLogConsumer(LoggerFactory.getLogger(AppContainer.class)));
        waitingFor(Wait.forHttp("/system/readiness").forPort(8085).forStatusCode(200));
    }

    public int getPort() {
        return getMappedPort(8080);
    }

    public URI getURI() {
        return URI.create(String.format("http://%s:%s", getHost(), getPort()));
    }
}

File: kora-kotlin-helloworld/src/test/kotlin/ru/tinkoff/kora/kotlin/example/helloworld/AppContainer.kt
package ru.tinkoff.kora.kotlin.example.helloworld

import org.slf4j.LoggerFactory
import org.testcontainers.containers.GenericContainer
import org.testcontainers.containers.output.Slf4jLogConsumer
import org.testcontainers.containers.wait.strategy.Wait
import org.testcontainers.images.builder.ImageFromDockerfile
import org.testcontainers.utility.DockerImageName
import java.net.URI
import java.nio.file.Paths
import java.util.concurrent.Future


class AppContainer : GenericContainer<AppContainer> {

    constructor(dockerImageName: DockerImageName) : super(dockerImageName)

    constructor(image: Future<String>) : super(image)

    companion object {

        fun build(): AppContainer {
            val appImage = System.getenv("IMAGE_KORA_KOTLIN_HELLOWORLD")
            return if (appImage != null && !appImage.isBlank())
                AppContainer(DockerImageName.parse(appImage))
            else AppContainer(
                ImageFromDockerfile("kora-kotlin-helloworld", true)
                    .withDockerfile(Paths.get("Dockerfile").toAbsolutePath())
            )
        }
    }

    override fun configure() {
        super.configure()
        withExposedPorts(8080, 8085)
        withLogConsumer(Slf4jLogConsumer(LoggerFactory.getLogger(AppContainer::class.java)))
        waitingFor(Wait.forHttp("/system/readiness").forPort(8085).forStatusCode(200))
    }

    val port: Int get() = getMappedPort(8080)

    val uri: URI get() = URI.create("http://$host:$port")
}

